### PR TITLE
Vertex tool: right-click to loop through editable features

### PR DESCRIPTION
**Pull request by wonder-sk**
Continuation of usability fixes in the vertex tool...

Until now mouse right-click could only select and deselect the highlighted feature to "lock" vertex tool (and numerical editor) to it, so that it is easier to focus only on editing of the particular feature. It was however still difficult to pick the right feature in case there were multiple features in one location or very close to each other. This is now solved by the fact that repeated right button clicks will loop through the editable features. So if there are two features in one location (A, B) then repeated right-clicks will select: A - B - nothing - A - B - nothing ...

peek 2019-02-04 22-16

@nirvn @bstroebl As usual your feedback is highly appreciated :)

### CODE REVIEW

File: src/app/vertextool/qgsvertextool.cpp [hunk 1]
```diff
@@ -849,6 +841,126 @@ QgsPointLocator::Match QgsVertexTool::snapToPolygonInterior( QgsMapMouseEvent *e
...
+QSet<QPair<QgsVectorLayer *, QgsFeatureId> > QgsVertexTool::findAllEditableFeatures( const QgsPointXY &mapPoint )
+{
+  QSet< QPair<QgsVectorLayer *, QgsFeatureId> > alternatives;
+
+  // if there is a current layer, it should have priority over other layers
+  // because sometimes there may be match from multiple layers at one location
+  // and selecting current layer is an easy way for the user to prioritize a layer
+  if ( QgsVectorLayer *currentVlayer = currentVectorLayer() )
+  {
+    for ( const QgsPointLocator::Match &m : findEditableLayerMatches( mapPoint, currentVlayer ) )
```
Comment (shahzadlone) [inline src/app/vertextool/qgsvertextool.cpp]: could you do ```const auto &m```

File: src/app/vertextool/qgsvertextool.cpp [hunk 2]
```diff
@@ -849,6 +841,126 @@ QgsPointLocator::Match QgsVertexTool::snapToPolygonInterior( QgsMapMouseEvent *e
...
+
+  QgsSnappingUtils *snapUtils = canvas()->snappingUtils();
+  QgsPointLocator *locator = snapUtils->locatorForLayer( layer );
+
+  if ( layer->geometryType() == QgsWkbTypes::PolygonGeometry )
+  {
+    matchList << locator->pointInPolygon( mapPoint );
+  }
+
+  double tolerance = QgsTolerance::vertexSearchRadius( canvas()->mapSettings() );
```
Comment (shahzadlone) [inline src/app/vertextool/qgsvertextool.cpp]: Does this get mutated ? if not we could make this ```const```.

File: src/app/vertextool/qgsvertextool.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/vertextool/qgsvertextool.cpp]: There was discussion on this when we first moved to C++11, but the outcome was to stick to explicit types UNLESS they are iterators or over complex template based types. (Avoiding auto improves readability)

**Comment by nyalldawson** [comment]
@wonder-sk 

>  I think when there is a selected (locked) feature, we should stop highlights of any other features so that the user can focus on that feature without getting distracted...?

+1 to this

File: src/app/vertextool/qgsvertextool.cpp (hunk 1 above)
Comment (shahzadlone) [reply src/app/vertextool/qgsvertextool.cpp]: That's a fair point, thanks for the info!

**Comment by nyalldawson** [comment]
@wonder-sk 

One oddity I've noticed with master: if I right click to lock a polygon, I can still drag to select vertices from a different polygon. These get highlighted in blue, but can't be moved/edited in any way. This only happens if the drag selection is entirely outside of the locked polygon. I think in this case nothing should be selected.

**Comment by nirvn** [comment]
Right click dragging shouldn't select vertices in any case, IMHO.

**Comment by nirvn** [comment]
+1 to merge this fix, and backport any vertex stuff that hasn't been committed to 3.4 prior to its next release. Great job uplifting this crucial tool.

**Comment by wonder-sk** [comment]
@nyalldawson agreed that when a feature is locked, it should not be possible to select vertices from other features. Let me address that in a follow up PR together with highlight.

@DelazJ I would create a ticket in qgis docs repo afterwards with a summary of all relevant changes since some things are still a subject to change based on testing

**Comment by nyalldawson** [comment]
@wonder-sk another change I think is worthwhile to make would be to change the default vertex tool to the single layer one. I remember this being discussed on the mailing list but I don't believe the change was ever made.

**Comment by bstroebl** [comment]
@nyalldawson : you are right, +1

**Comment by andreasneumann** [comment]
> @wonder-sk another change I think is worthwhile to make would be to change the default vertex tool to the single layer one. I remember this being discussed on the mailing list but I don't believe the change was ever made.

I would appreciate that as well. It helps to stay "safe" without too many unwanted side effects on other layers. And it is also what existing users are already used to, from previous QGIS versions.

If someone wants to edit in multiple layers at once, he/she should consciously/deliberately do this by selecting the multi-layer editing.

**Comment by bstroebl** [comment]
that reminds me of our diskussion via email, where I explained my understanding about topology and multi-layer mode as 

> the vertex tool's mode (current/all) decides on which layers are changed, topological editing decides on which features (only selected or all that share this node) are changed.

Is this agreed upon? If yes we should think about what should happen if the mode is all layers but topological editing is off: the node to be moved could be part of two features in another layer:

1. choose a feature arbitrarily?
2. automatically enable topolgical editing (-1 from me)
3. inform user

**Comment by wonder-sk** [comment]
Okay - let's switch to "current layer" mode as the default.

> the vertex tool's mode (current/all) decides on which layers are changed, topological editing decides on which features (only selected or all that share this node) are changed.

@bstroebl Yes I think we are in agreement here. Regarding your other question, what should happen in "all layers" mode with topo editing off... I don't think we need some special handling for this case - vertex tool mode and topo editing are independent options and all of their combinations make sense to me. So it is not clear to me why do you see a need for special handling of "all layers" mode with topo editing enabled. Maybe an example would help (with what you expect vs what happens).

**Comment by bstroebl** [comment]
@wonder-sk Yes, both options are independent and will normally work, my thinking was the following scenario:
two layers in edit mode, layer A is the current layer
_all layers_ is **on** and _topolgical editing_ is **off** 
the vertex to be moved is a vertex in more than one feature in layer B
=> all layers mode tells us to move identical vertices in both, A and B
=> topological editing off tells us to not move a vertex as a common vertex 
I cannot tell what is currently happening and above I have listed three solutions what might happen

**Comment by bstroebl** [comment]
@nirvn I can only lock features in the current layer, my question is which feature shall be changed in **another** editable layer if all-layers mode is on and a vertex of the locked feature in the current layer is a vertex in e.g. two features in another layer (and topo mode is off).

**Comment by nirvn** [comment]
Oh that might be my wrong here, I thought locking onto a feature worked on all layers (assuming you are using the all layers mode).

@wonder-sk , can a right click cycle through all overlapping features from all edit-enabled layers when using the all layers mode?

**Comment by bstroebl** [comment]
Oh, could be my fault, too. I actually **thought** that one could only lock features in the current layer, but obviuosly this is not true. I will do furterh tests and call back.

**Comment by nirvn** [comment]
My understanding is that it'll cycle through all features.

The culmination of all recent changes as dealt with a lot of these corner scenarios. Happy times.

**Comment by bstroebl** [comment]
Forget it, it was me missunderstanding what _all layers_ actually means. I thought you can only edit features in the current layer and if _all layers_ is active the same vertex in other layers is edited, too. Now I see that _all layers_ means you can edit features in all editable layers, not just within the current layer. Common vertices are moved together if _topological editing_ is enabled. 

However I feature the following problem: When locking a feature in the not current layer I cannot left-click (in order to move) a vertex that is a common vertex in a feature of the current layer. I can left-click on vertices that are only in the currently locked feature. Topo settings have no influence.
I tested with code version https://github.com/qgis/QGIS/commit/4a44306700

**Comment by wonder-sk** [comment]
> Although I do not see any obstacles now anymore, I would expect other users to be as surprised as me when features are changed in other than the current layer although the vertex-tool mode is _current layer_.

I see what you mean now... your understanding of "current layer" mode is that only the current layer may be modified, while the current implementation of "current layer" mode is that only the current layer may be interacted with, but with topo editing on also other layers may get modified. I believe that's what you reported in the following ticket, right? https://issues.qgis.org/issues/20505

I do not have a strong opinion how this should work - I can see both options valid...

**Comment by bstroebl** [comment]
>  I believe that's what you reported in the following ticket, right? https://issues.qgis.org/issues/20505

Exactly; it can be implemented either way and there might always be someone that would expect the tool's behaviour to be the other way round :-)
I doubt we can find words for the tool tip that describe the way the tool is working unambiguously (we should try, though).

**Comment by bstroebl** [comment]
@wonder-sk This seems to only work when moving the mouse pointer in between clicks in areas where several features overlap. If I only do consecutive right clicks at the same position nothing happens. Same when trying to lock a feature in the first place: Move mouse pointer into overlapping area => one feature is highlighted - right click => no feature is locked - move mouse and right click => highlighted feature is locked. Behaviour is different in non-overlapping area. There the highlighted feature is immediately locked when I click right.
Tested on Win7

**Pull request merged**